### PR TITLE
"Recipe tagging" feature addition

### PR DESCRIPTION
**Pull request by th-bunratta**
This will introduce changes as follows:

- Users can now **tag** recipes and assign labels to them with a dedicated widget. For instance,
  - **Allergy** | Gluten
  - **Allergy** | Shellfish
  - **Vegetarian**

- User can type in recipe tag widget equipped with autocomplete assistance from the existing tags in earlier recipes

P.S. Other changes are about to be incrementally posted here.

### CODE REVIEW

File: koocook_core/models/post.py [hunk 1]
```diff
@@ -21,7 +21,8 @@ class Post(SerialisableModel, ReviewableModel, models.Model):
     aggregate_rating = models.OneToOneField(
         'koocook_core.AggregateRating',
         on_delete=models.PROTECT,
-        blank=True
+        blank=True, null=True,
+        # default=create_empty_aggregate_rating,
     )
```
Comment (MaiMee1) [inline koocook_core/models/post.py]: Why do you need this to be nullable?

File: koocook_core/models/recipe.py [hunk 2]
```diff
@@ -33,7 +33,7 @@ class Recipe(ReviewableModel, models.Model):
     aggregate_rating = models.OneToOneField(
         'koocook_core.AggregateRating',
         on_delete=models.PROTECT,
-        blank=True,
+        blank=True, null=True
     )
```
Comment (MaiMee1) [inline koocook_core/models/recipe.py]: This as well, I thought you had it auto-create on saving?

File: static/js/bundle.js [hunk 3]
```diff
@@ -0,0 +1 @@
...t){return n.call(t)}},function(t,e,n){var r=n(41),o=Object.prototype,i=o.hasOwnProperty,a=o.toString,s=r?r.toStringTag:void 0;t.exports=function(t){var e=i.call(t,s),n=t[s];try{t[s]=void 0;var r=!0}catch(t){}var o=a.call(t);return r&&(e?t[s]=n:delete t[s]),o}},function(t,e,n){var r=n(41),o=n(79),i=n(78),a="[object Null]",s="[object Undefined]",u=r?r.toStringTag:void 0;t.exports=function(t){return null==t?void 0===t?s:a:u&&u in Object(t)?o(t):i(t)}},function(t,e,n){var r=n(80),o=n(77),i="[object Symbol]";t.exports=function(t){return"symbol"==typeof t||o(t)&&r(t)==i}},function(t,e,n){var r=n(43),o=n(81),i=NaN,a=/^\s+|\s+$/g,s=/^[-+]0x[0-9a-f]+$/i,u=/^0b[01]+$/i,c=/^0o[0-7]+$/i,l=parseInt;t.exports=function(t){if("number"==typeof t)return t;if(o(t))return i;if(r(t)){var e="function"==typeof t.valueOf?t.valueOf():t;t=r(e)?e+"":e}if("string"!=typeof t)return 0===t?t:+t;t=t.replace(a,"");var n=u.test(t);return n||c.test(t)?l(t.slice(2),n?2:8):s.test(t)?i:+t}},function(t,e){var n;n=function(){return this}();try{n=n||Function("return this")()||(0,eval)("this")}catch(t){"object"==typeof window&&(n=window)}t.exports=n},function(t,e,n){(function(e){var n="object"==typeof e&&e&&e.Object===Obje
```
Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: [Found fs.open with non literal argument at index 0 (security/detect-non-literal-fs-filename)](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

**Comment by KooCook** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 31
           

Complexity increasing per file
==============================
- koocook_core/views/posts.py  1
- koocook_core/views/mixins.py  10
- koocook_core/views/recipes.py  1
- koocook_core/templatetags/recipe_extras.py  1
- static/js/bundle.js  21
- koocook_core/tests/test_recipe.py  2
- koocook_core/tests/test_templatetags/test_custom_filters.py  1
         

Clones added
============
- koocook_core/tests/test_post.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/koocook-dj/koocook-dj/pullRequest?prid=4590732&bid=15414227)

File: koocook_core/views/forms/recipe.py [hunk 4]
```diff
@@ -1,9 +1,11 @@
 from django import forms
+from django.forms.widgets import HiddenInput
```
Comment (KooCook) [inline koocook_core/views/forms/recipe.py]: Codacy Issue found: ['django.forms.widgets.HiddenInput' imported but unused](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

File: koocook_core/tests/test_recipe.py [hunk 5]
```diff
@@ -1,10 +1,54 @@
...
+
+    def test_recipe_create_view(self):
+        with self.subTest("Authenticated user access"):
+            response = self.client.get(reverse("koocook_core:recipe-create"))
+            self.assertTemplateUsed(response, "recipes/create.html")
+        with self.subTest("Unauthenticated user access"):
+            self.client.logout()
+            response = self.client.get(reverse("koocook_core:recipe-create"))
+            self.assertRedirects(response,
+                                 f"{reverse('social:begin', args=['google-oauth2'])}?next={reverse('koocook_core:recipe-create')}",
```
Comment (KooCook) [inline koocook_core/tests/test_recipe.py]: Codacy Issue found: [invalid syntax](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

File: static/js/bundle.js (hunk 3 above)
Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: [Script URL is a form of eval. (no-script-url)](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: ['U' is a function. (no-func-assign)](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

File: koocook_core/views/forms/recipe.py (hunk 4 above)
Comment (KooCook) [inline koocook_core/views/forms/recipe.py]: Codacy Issue found: [Unused HiddenInput imported from django.forms.widgets](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

File: static/js/bundle.js (hunk 3 above)
Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: ['Vue' is not defined. (no-undef)](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: [Expected parentheses around arrow function argument. (arrow-parens)](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: [Method 'onFreeze' has a complexity of 5. (complexity)](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: [Use ===/!== to compare with true/false or Numbers](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: [Strings must use doublequote. (quotes)](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: [Expected a 'break' statement before 'case'. (no-fallthrough)](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

File: koocook_core/views/handlers.py [hunk 6]
```diff
@@ -1,9 +1,11 @@
+import json
```
Comment (KooCook) [inline koocook_core/views/handlers.py]: Codacy Issue found: [Unused import json](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

File: static/js/bundle.js (hunk 3 above)
Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: [eval can be harmful. (no-eval)](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: [Generic Object Injection Sink (security/detect-object-injection)](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: [Expected '!==' and instead saw '!='. (eqeqeq)](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

File: koocook_core/views/handlers.py (hunk 6 above)
Comment (KooCook) [inline koocook_core/views/handlers.py]: Codacy Issue found: ['json' imported but unused](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

File: static/js/bundle.js (hunk 3 above)
Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: ['_' used outside of binding context. (block-scoped-var)](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: [Found non-literal argument to RegExp Constructor (security/detect-non-literal-regexp)](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: [Avoid assignments in operands](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: [Expected method shorthand. (object-shorthand)](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: [Missing semicolon. (semi)](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: [Expected a conditional expression and instead saw an assignment. (no-cond-assign)](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: [A constructor name should not start with a lowercase letter. (new-cap)](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: [Missing space after =>. (arrow-spacing)](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

**Comment by KooCook** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 32
           

Complexity increasing per file
==============================
- koocook_core/templatetags/recipe_extras.py  1
- static/js/bundle.js  21
- koocook_core/views/mixins.py  10
- koocook_core/views/posts.py  1
- koocook_core/tests/test_templatetags/test_custom_filters.py  1
         

Clones added
============
- koocook_core/tests/test_post.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/koocook-dj/koocook-dj/pullRequest?prid=4590732&bid=15414227)

File: static/js/bundle.js (hunk 3 above)
Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: [Avoid assignments in operands](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: ['U' is a function. (no-func-assign)](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: [Strings must use doublequote. (quotes)](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: [Expected { after 'if' condition. (curly)](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

File: koocook_core/views/forms/recipe.py (hunk 4 above)
Comment (KooCook) [inline koocook_core/views/forms/recipe.py]: Codacy Issue found: ['django.forms.widgets.HiddenInput' imported but unused](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

File: static/js/bundle.js (hunk 3 above)
Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: [Script URL is a form of eval. (no-script-url)](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: [Expected method shorthand. (object-shorthand)](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: [Expected a 'break' statement before 'case'. (no-fallthrough)](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

File: koocook_core/views/mixins.py [hunk 7]
```diff
@@ -53,7 +53,33 @@ def get_context_data(self, **kwargs):
 
 class RecipeViewMixin:
     def form_valid(self, form):
+        from ..models import Tag, TagLabel
         response = super().form_valid(form)
+        tags = json.loads(self.request.POST.get('tags'))
+        if tags:
+            for tag in tags:
+                tag_body: dict = {field: tag[field] for field in tag if field
```
Comment (KooCook) [inline koocook_core/views/mixins.py]: Codacy Issue found: [invalid syntax](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

File: static/js/bundle.js (hunk 3 above)
Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: ['Vue' is not defined. (no-undef)](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: [Use ===/!== to compare with true/false or Numbers](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: [Method 'onFreeze' has a complexity of 5. (complexity)](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: [Found non-literal argument to RegExp Constructor (security/detect-non-literal-regexp)](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: [The numeric literal '4294967295' will have at different value at runtime.](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

File: koocook_core/views/handlers.py (hunk 6 above)
Comment (KooCook) [inline koocook_core/views/handlers.py]: Codacy Issue found: [Unused import json](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

File: static/js/bundle.js (hunk 3 above)
Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: ['_' used outside of binding context. (block-scoped-var)](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

File: koocook_core/views/handlers.py (hunk 6 above)
Comment (KooCook) [inline koocook_core/views/handlers.py]: Codacy Issue found: ['json' imported but unused](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

File: static/js/bundle.js (hunk 3 above)
Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: [Empty block statement. (no-empty)](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: [Expected '!==' and instead saw '!='. (eqeqeq)](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: [Found fs.open with non literal argument at index 0 (security/detect-non-literal-fs-filename)](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: [The body of a for-in should be wrapped in an if statement to filter unwanted properties from the prototype. (guard-for-in)](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

File: koocook_core/views/forms/recipe.py (hunk 4 above)
Comment (KooCook) [inline koocook_core/views/forms/recipe.py]: Codacy Issue found: [Unused HiddenInput imported from django.forms.widgets](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

File: static/js/bundle.js (hunk 3 above)
Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: [Expected parentheses around arrow function argument. (arrow-parens)](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: ['v' was used before it was defined. (no-use-before-define)](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

Comment (KooCook) [inline static/js/bundle.js]: Codacy Issue found: [Wrap an immediate function invocation in parentheses. (wrap-iife)](https://app.codacy.com/app/KooCook/koocook-dj/pullRequest?prid=4590732)

**Pull request merged**